File: js_implement/sort/shellSort.ts
export const shellSort = <T>(arr: T[], n: number) => {
    let h = 1
    while (h < n / 3) {
        h = 3 * h + 1
    }
    while (h >= 1) {
        for (let i = h; i < n; i++) {
            let temp: T = arr[i]
            let j: number
            for (j = i; j >= h && arr[j - h] > temp; j -= h) {
                arr[j] = arr[j - h]
            }
            arr[j] = temp
        }
        h = Math.floor(h / 3)
    }
}

File: rust_implement/src/sort/shell_sort.rs
pub fn shell_sort(arr: &mut Vec<i32>) {
    let n = arr.len();
    let mut h = 1;
    while h < n / 3 {
        h = 3 * h + 1;
    }
    while h >= 1 {
        for i in h..n {
            let e = arr[i];
            let mut j = i;
            while j >= h && arr[j - h] > e {
                arr[j] = arr[j - h];
                j -= h;
            }
            arr[j] = e;
        }
        h = h / 3;
    }
}
